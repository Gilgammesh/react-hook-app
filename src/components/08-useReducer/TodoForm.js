import React from "react";
import PropTypes from "prop-types";
import { useForm } from "../../hooks/useForm";

const TodoForm = ({ dispatch }) => {
  const initialForm = {
    id: "",
    desc: "",
    done: false,
  };
  const [formValues, handleInputChange, resetForm] = useForm(initialForm);
  const { desc, done } = formValues;

  const handleSubmit = (evt) => {
    evt.preventDefault();
    if (!desc.trim().length) {
      return;
    }
    dispatch({
      type: "add",
      payload: { ...formValues, id: new Date().getTime() },
    });
    resetForm(initialForm);
  };

  return (
    <form onSubmit={handleSubmit}>
      <h3>Crear nueva tarea</h3>
      <div className="form-group">
        <label htmlFor="descripcion">Descripcion: </label>
        <input
          name="desc"
          type="text"
          className="form-control"
          id="descripcion"
          value={desc}
          placeholder="Ingrese descripcion de tarea"
          onChange={handleInputChange}
        />
      </div>
      <div className="form-check">
        <input
          name="done"
          type="checkbox"
          className="form-check-input"
          id="done"
          checked={done}
          onChange={handleInputChange}
        />
        <label className="form-check-label" htmlFor="done">
          Realizado
        </label>
      </div>
      <button type="submit" className="btn btn-primary mt-3">
        AÃ±adir
      </button>
    </form>
  );
};

TodoForm.propTypes = {
  dispatch: PropTypes.func.isRequired,
};

export default TodoForm;
