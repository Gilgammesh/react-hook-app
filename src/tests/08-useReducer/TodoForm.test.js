import React from "react";
import { shallow } from "enzyme";
import "@testing-library/jest-dom";
import TodoForm from "../../components/08-useReducer/TodoForm";

describe("Pruebas unitarias en el componente TodoForm.js", () => {
  const dispatch = jest.fn();
  const func = () => {};
  /************************************************************************************/
  test("se debe mostrar el componente <TodoForm /> correctamente", () => {
    const wrapper = shallow(<TodoForm dispatch={dispatch} />);
    expect(wrapper).toMatchSnapshot();
  });
  /************************************************************************************/
  test("no se debe de llamar la funcion dispatch, mientras la desc este en blanco", () => {
    const wrapper = shallow(<TodoForm dispatch={dispatch} />);
    const form = wrapper.find("form");
    form.simulate("submit", { preventDefault: func });
    expect(dispatch).toHaveBeenCalledTimes(0);
    expect(dispatch).not.toHaveBeenCalled();
  });
  /************************************************************************************/
  test("se debe de llamar la funcion dispatch al cambiar el input descripciÃ³n", () => {
    const value = "Aprender React";
    const name = "desc";
    const type = "text";
    const checked = false;
    const wrapper = shallow(<TodoForm dispatch={dispatch} />);
    const handleInputChange = wrapper
      .find("input#descripcion")
      .prop("onChange");
    handleInputChange({ target: { type, checked, name, value } });
    const form = wrapper.find("form");
    form.simulate("submit", { preventDefault: func });
    expect(dispatch).toHaveBeenCalled();
    expect(dispatch).toHaveBeenCalledTimes(1);
    expect(dispatch).toHaveBeenCalledWith({
      payload: {
        desc: value,
        done: false,
        id: expect.any(Number),
      },
      type: "add",
    });
    const input = wrapper.find("input#descripcion");
    expect(input.prop("value").trim()).toBe("");
  });
});
