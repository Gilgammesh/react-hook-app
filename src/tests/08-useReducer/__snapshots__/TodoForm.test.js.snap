// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas unitarias en el componente TodoForm.js se debe de llamar la funcion dispatch 1`] = `
<form
  onSubmit={[Function]}
>
  <h3>
    Crear nueva tarea
  </h3>
  <div
    className="form-group"
  >
    <label
      htmlFor="descripcion"
    >
      Descripcion: 
    </label>
    <input
      className="form-control"
      id="descripcion"
      name="desc"
      onChange={[Function]}
      placeholder="Ingrese descripcion de tarea"
      type="text"
      value=""
    />
  </div>
  <div
    className="form-check"
  >
    <input
      checked={false}
      className="form-check-input"
      id="done"
      name="done"
      onChange={[Function]}
      type="checkbox"
    />
    <label
      className="form-check-label"
      htmlFor="done"
    >
      Realizado
    </label>
  </div>
  <button
    className="btn btn-primary mt-3"
    type="submit"
  >
    Añadir
  </button>
</form>
`;

exports[`Pruebas unitarias en el componente TodoForm.js se debe mostrar el componente <TodoForm /> correctamente 1`] = `
<form
  onSubmit={[Function]}
>
  <h3>
    Crear nueva tarea
  </h3>
  <div
    className="form-group"
  >
    <label
      htmlFor="descripcion"
    >
      Descripcion: 
    </label>
    <input
      className="form-control"
      id="descripcion"
      name="desc"
      onChange={[Function]}
      placeholder="Ingrese descripcion de tarea"
      type="text"
      value=""
    />
  </div>
  <div
    className="form-check"
  >
    <input
      checked={false}
      className="form-check-input"
      id="done"
      name="done"
      onChange={[Function]}
      type="checkbox"
    />
    <label
      className="form-check-label"
      htmlFor="done"
    >
      Realizado
    </label>
  </div>
  <button
    className="btn btn-primary mt-3"
    type="submit"
  >
    Añadir
  </button>
</form>
`;
